Improve Prettier, formatting enforcement

sammoore opened the pull request:
This PR updates prettier, runs it on the pre-commit hook, and updates the VSCode/editorconfig files to ensure settings are used regardless of the workspace used in the editor.

- [x] Prettier: Use [`lint-staged`](https://github.com/lint-staged/lint-staged/) to run Prettier on the relevant committed files upon `git commit`.
  * The previous Husky commit hook was utilizing [`pretty-quick`](https://github.com/azz/pretty-quick), which hasn't been updated in about 18 months and [does not work with Prettier v3](https://github.com/azz/pretty-quick/issues/168), which the majority of this project does/did utilize (see next bullet). 
- [x] Move VSCode settings (i.e. from `cli/.vscode/settings.json`) to `.vscode/settings.json`
- [x] Create symlinks from `{cli,docs,www}/.vscode/settings.json` to the top level settings.json, to ensure the settings are used regardless of the folder that's opened as the workspace.
- [x] Add `.vscode/extensions.json` for recommended/unwanted extensions, and use similar symlinks as above for sub-projects.
- [x] Added `.editorconfig` for generic editor support for line endings and tab style/sizing.

danstepanov reviewed (commented):
Will want to get those build steps to succeed before merging

sammoore commented on .gitignore:
```diff
@@ -8,4 +8,3 @@ build
 .vscode
 myTestProject
 .turbo
-package-lock.json
```
Per the commit message in `b125c38`

> Using bun, not npm, so package-lock.json should not exist

danstepanov replied:
I don't recall where we landed with this conversation yesterday. All I know is that vercel throws an error in the build step if a `package-lock.json` is present.

sammoore replied:
That's right -- removing it from .gitignore **_doesn't_** mean we want to commit it. There should never be one, unless someone ran `npm install` and then added & committed the file. (And if you ran `npm install`, it's possible to write code that compiles on your machine but will fail on others as modules could be resolved to different versions).

I suggested that it's useful to see `package-lock.json` as having been created when using `git status`/etc, in order to indicate that you did something wrong, but I think we landed that you'd rather not tell potential contributors to remove `package-lock.json` if they accidentally commit it, so to leave it in the `.gitignore`.

sammoore replied:
Reverted the commit, resolving conversation.

sammoore commented on .lintstagedrc.js:
```diff
@@ -0,0 +1,6 @@
+/**
```
This is the configuration for [`lint-staged`](https://github.com/lint-staged/lint-staged/), so that it runs `prettier` upon commit, only on the committed/modified files. It replaces `pretty-quick`, but since `lint-staged` is a more general tool, this configuration specifies that it should run `prettier`.

sammoore commented on cli/src/templates/packages/expo-router/stack/app/index.tsx.ejs:
```diff
@@ -15,7 +15,7 @@ export default function Page() {
 				<View className={styles.main}>
 				<View>
-					<Text className={styles.title}>Hello World</Text>
+					<Text className={styles.title}>Hello Test</Text>
```
It was a mistake to include this in a commit, and was just testing a modification of `cli/` while working on this commit. I will remove this 👍

sammoore commented:
Rebased and updated with my latest changes and with the builds resolved (I think by upgrading bun and then updating the lockb file?). 

A few pending things:

- [ ] Top-level `package.json`'s `format` run-script could be more exhaustive & [based on my findings](https://discord.com/channels/1173879003191459860/1175505640156844084/1182458159617556540) while creating the `.prettierignore`.
- [ ] [This commit needs a different name or to be removed](https://github.com/danstepanov/create-expo-stack/pull/102/commits/b041b32116fd0a867f7d82e48ac9fb3e65a17c89)
  * I accidentally gave it another name during some sort of commit cleanup.
  * It might not be necessary anymore, it was an attempt to fix the build process in CI but I don't think it was the solution.
- [ ] PR has a lot of separate but related changes now; might be worth separating them?

danstepanov commented:
@sammoore yea can you break the PR up if it addresses different things?

sammoore commented:
"Half" of this PR (Husky fixes & the automated `bun run changeset`) is now in #125 

After that's merged, will rebase and update this PR with just the Prettier/tabSize/etc changes.

sammoore commented:
For the curious: updating the PR with rebased commits (against the other portion of this original PR that have been merged in #125).

Still need to update description, verify locally, and check tests.

sammoore commented:
I've moved the `cli/` prettier version bump to #126 since (a) I'm able to verify the changes easier and (b) it simplifies this PR, removing changes unrelated to the core description.

I've updated the description of this PR, but it has yet to be rebased until #126 is merged.

sammoore commented:
Rebased against #126 and #125. Still need to verify changes locally ([empty checkbox here](https://github.com/danstepanov/create-expo-stack/pull/102#issuecomment-1859635775)) and make sure nothing else is needed.

sammoore commented:
Removed the change from tabs to 2 spaces. Will include in the PR that actually applies the switch.

Also, updated the PR description to mention `.editorconfig`.

Rebased and ready for review.

danstepanov reviewed (approved):
LGTM
